File: Array&Hashing/3346. Maximum Frequency of an Element After Performing Operations I/solution.py
class Solution:
    def maxFrequency(self, nums: List[int], k: int, numOperations: int) -> int:
        mn = min(nums)
        mx = max(nums)
        diff = defaultdict(int)
        for num in nums:
            diff[num-k] += 1
            diff[num+k+1] -= 1
        
        for i in range(mn-k, mx+k+1):
            diff[i] += diff[i-1]

        count = Counter(nums)
        res = 0
        for i in range(mn, mx+1):
            res = max(res, count[i] + min(numOperations, diff[i]-count[i]))
        return res
        
# - use difference array to know the maximum frequency in range [min(nums)-k, max(nums)+k]
# - iterate value in range of [min(nums), max(nums)] to find maximum possible frequency of valid value

File: Array&Hashing/3346. Maximum Frequency of an Element After Performing Operations I/solution.cpp
class Solution {
public:
    int maxFrequency(vector<int>& nums, int k, int numOperations) {
        int n = nums.size();

        unordered_map<int,int> count;
        unordered_map<int,int> diff;
        for (auto& num : nums) {
            diff[num-k]++;
            diff[num+k+1]--;
            
            count[num]++;
        }
        
        
        int mn = *min_element(nums.begin(), nums.end()), mx = *max_element(nums.begin(), nums.end());
        for (int i=mn-k; i<mx+k+1; i++) {
            diff[i] += diff[i-1];
        }

        int res = 0;
        for (int num=mn; num<=mx; num++) {
            int op = diff[num] - count[num];
            res = max(res, count[num] + min(numOperations, op));
        }        

        return res;
    }
};